app/api/chat/route.ts

```diff
@@ -0,0 +1,119 @@
+import { StreamingTextResponse, LangChainStream, Message } from "ai";
+import { ChatOpenAI } from "langchain/chat_models/openai";
+import {
+  AIChatMessage,
+  HumanChatMessage,
+  SystemChatMessage,
+} from "langchain/schema";
+import { kv } from "@vercel/kv";
+import { getSalesforceDescriptions } from "./getSalesforceDescriptions";
+import { NextResponse } from "next/server";
+import {
+  CachedSalesforceData,
+  CustomFieldsOnObject,
+  CustomObject,
+  SalesforceAuthCache,
+} from "@/shared/types/salesforceTypes";
+import * as process from "process";
+
+export const runtime = "edge";
+
+export async function POST(req: Request) {
+  const { messages, salesforceId } = await req.json();
+  const cachedRes: SalesforceAuthCache | null = await kv.get(salesforceId);
+  if (!cachedRes) throw Error("No salesforce token");
+
+  let cachedSalesforceData = (await kv.get(`${salesforceId}data`)) as string;
+
+  let salesforceData: CachedSalesforceData;
+  if (!cachedSalesforceData) {
+    cachedSalesforceData = (await getSalesforceDescriptions(
+      cachedRes?.accessToken as string,
+      cachedRes?.refreshToken as string,
+      cachedRes?.instanceUrl as string,
+      salesforceId as string,
+    )) as string;
+
+    await kv.set(`${salesforceId}data`, cachedSalesforceData as string);
+  }
+
+  if (!cachedSalesforceData) {
+    throw Error("No salesforce data even after regenerating access token");
+  }
+
+  const { stream, handlers } = LangChainStream();
+
+  // check count
+  const count: number = (await kv.get(`${salesforceId}count`)) as number;
+  if (count >= 50 && salesforceId !== process.env.CUE_SALESFORCE_ID) {
+    return new NextResponse(
+      "You seem to have exceeded your free trial. Please contact christina@trycue.ai to ask me more questions.",
+    );
+  }
+  if (typeof count === "undefined") throw Error("No count for some reason");
+  await kv.set(`${salesforceId}count`, (count as number) + 1);
```

Comment:
For your kv fields you probably want to include a `/` deliminator so it's easier for us to search the keys available

Reply:
i.e. `${salesforceId}/count`